pkg/referrerstore/oras/oras.go

```diff
@@ -157,62 +169,108 @@ func (store *orasStore) ListReferrers(ctx context.Context, subjectReference comm
+	var err error
+	repository, err := store.createRepository(ctx, subjectReference)
 	if err != nil {
 		return nil, err
 	}
 
-	ref := fmt.Sprintf("%s@%s", subjectReference.Path, digest)
-	desc, err := oras.Copy(ctx, registryClient, ref, store.localCache, "")
+	// resolve blob descriptor if not provided
+	var resolvedBlobDesc oci.Descriptor
+	if len(blobDesc) > 0 {
+		resolvedBlobDesc = blobDesc[0]
+	} else {
+		ref := fmt.Sprintf("%s@%s", subjectReference.Path, digest)
+		resolvedBlobDesc, err = repository.Blobs().Resolve(ctx, ref)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// check if blob exists in local ORAS cache
+	isCached, err := store.localCache.Exists(ctx, resolvedBlobDesc)
 	if err != nil {
 		return nil, err
 	}
 
-	return store.getRawContentFromCache(ctx, desc)
+	if !isCached {
+		// fetch blob content from remote repository
+		rc, err := repository.Fetch(ctx, resolvedBlobDesc)
+		if err != nil {
+			return nil, err
+		}
+
+		// push fetched content to local ORAS cache
+		err = store.localCache.Push(ctx, resolvedBlobDesc, rc)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return store.getRawContentFromCache(ctx, resolvedBlobDesc)
 }
 
 func (store *orasStore) GetReferenceManifest(ctx context.Context, subjectReference common.Reference, referenceDesc ocispecs.ReferenceDescriptor) (ocispecs.ReferenceManifest, error) {
-	client, err := store.createRegistryClient(ctx, subjectReference)
+	repository, err := store.createRepository(ctx, subjectReference)
 	if err != nil {
 		return ocispecs.ReferenceManifest{}, err
 	}
-
-	var result ocispecs.ReferenceManifest
-	artifactManifestFound := false
-	_, err = oras.Graph(ctx, subjectReference.Original, referenceDesc.ArtifactType, client.Resolver,
-		func(parent artifactspec.Descriptor, parentManifest artifactspec.Manifest, objects []target.Object) error {
-			if parent.Digest == referenceDesc.Digest {
-				result = ArtifactManifestToReferenceManifest(parentManifest)
-				artifactManifestFound = true
-			}
-			return nil
-		})
-
+	var manifestBytes []byte
+	// check if manifest exists in local ORAS cache
+	isCached, err := store.localCache.Exists(ctx, referenceDesc.Descriptor)
 	if err != nil {
 		return ocispecs.ReferenceManifest{}, err
 	}
 
-	if !artifactManifestFound {
-		return ocispecs.ReferenceManifest{}, fmt.Errorf("cannot find artifact manifest with digest %s", referenceDesc.Digest)
+	if !isCached {
+		// fetch manifest content from repository
+		manifestReader, err := repository.Fetch(ctx, referenceDesc.Descriptor)
+		if err != nil {
+			return ocispecs.ReferenceManifest{}, err
+		}
+
+		manifestBytes, err = io.ReadAll(manifestReader)
+		if err != nil {
+			return ocispecs.ReferenceManifest{}, err
+		}
+
+		// push fetched manifest to local ORAS cache
+		err = store.localCache.Push(ctx, referenceDesc.Descriptor, bytes.NewReader(manifestBytes))
```

Comment:
Can we pass  manifestReader to the cache instead of reading them at line 233

Reply:
the `manifestBytes` must be populated with the fetched Manifest bytes since it doesn't exist in the cache yet and we use it below. I could change it so that once it's added in the cache it pulls from cache again instead of just directly reading the reader returned by Fetch. but not sure what the value is since in any case the bytes would have to be read.